File: CPP/Sorting/lower_bound/Successful Pairs of Spells and Potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        vector<int>pairs;
        // -------------- will give TLE : T.C = O(N*M) ----------------
        /*for(int i=0;i<spells.size();i++){
            int cnt=0;
            for(int j=0;j<potions.size();j++){
                long long product=1LL*spells[i]*potions[j];
                if(product>=success)cnt++;
            }
            pairs.push_back(cnt);
        }*/

        sort(potions.begin(),potions.end());
        for(auto spell:spells){
            // minimum potion value needed for success
            long long need = ceil((double)success/spell);

            // find first potion >= need
            auto it=lower_bound(potions.begin(),potions.end(),need);
            pairs.push_back(potions.end()-it);
        }
        return pairs;        
    }
};

File: Java/Sorting/lowerBound/Successful Pairs of Spells and Potions.java
class Solution {
    // Custom lowerBound: first index with value >= target
    private int lowerBound(int[] arr,long need){
        int start=0;
        int end=arr.length;
        while(start<end){
            int mid=start+(end-start)/2;
            if(arr[mid]<need){
                start=mid+1;
            }else{
                end=mid;
            }
        }
        return start;
    }
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n=spells.length;
        int[] pairs=new int[n];
        Arrays.sort(potions);
        int i=0;
        for(int spell:spells){
            long need=(success + spell - 1) / spell;      //Math.ceil(success/spell);
            int idx=lowerBound(potions,need);
            pairs[i]=potions.length-idx;
            i++;
        }
        return pairs;
    }
}
